### PR TITLE
Bucket in path

### PR DESCRIPTION
**Pull request by hultin**

### CODE REVIEW

File: tests/django_s3_storage_test/settings.py [hunk 1]
```diff
@@ -16,28 +16,17 @@
...
 
-AWS_S3_BUCKET_NAME_STATIC = os.environ.get("AWS_S3_BUCKET_NAME")
-
-AWS_S3_KEY_PREFIX_STATIC = uuid.uuid4().hex
-
 DEFAULT_FILE_STORAGE = "django_s3_storage.storage.S3Storage"
 
-STATICFILES_STORAGE = "django_s3_storage.storage.ManifestStaticS3Storage"
-
+DEBUG = True
```
Comment (oskarerikmartinandersson) [inline tests/django_s3_storage_test/settings.py]: What implications does this have?

**Comment by oskarerikmartinandersson** [comment]
How do think about documentation in this repository? Should we updated it to match the code changes, or should we remove it so it's not out-of-sync with the code?

We also have other package/project related stuff we should maybe think about, e.g.
- Changelog
- License 
- Package related code (e.g. supported versions)
- Github related code

File: tests/django_s3_storage_test/settings.py [hunk 2]
```diff
@@ -46,6 +35,6 @@
 DATABASES = {
     "default": {
         "ENGINE": "django.db.backends.sqlite3",
-        "NAME": os.path.join(BASE_DIR, "db.sqlite3"),
+        "NAME": 'mydatabase',
```
Comment (oskarerikmartinandersson) [inline tests/django_s3_storage_test/settings.py]: Is there a reason to change?

File: django_s3_storage/storage.py
Comment (oskarerikmartinandersson) [inline django_s3_storage/storage.py:1]: I'm not sure on what order is best for classes, constants, functions etc. But right now everything seems a bit mixed, maybe move `Endpoints` down to the other classes?

File: django_s3_storage/storage.py [hunk 3]
```diff
@@ -92,34 +123,81 @@ def open(self, mode="rb"):
...
+    AWS_S3_ENCRYPT_KEY: bool = False
+    AWS_S3_KMS_ENCRYPTION_KEY_ID: str = ""
+    AWS_S3_GZIP: bool = True
+    AWS_S3_SIGNATURE_VERSION: str = "s3v4"
+    AWS_S3_FILE_OVERWRITE: bool = False
+    AWS_S3_USE_THREADS: bool = True
+    AWS_S3_MAX_POOL_CONNECTIONS: int = 10
+    AWS_S3_CONNECT_TIMEOUT: int = 60  # 60 seconds
+    AWS_S3_BUCKET_NAME: str = 'Deprecated'
+    AWS_S3_READ_ONLY: bool = False
```
Comment (oskarerikmartinandersson) [inline django_s3_storage/storage.py]: Keep order from original settings, i.e. move `AWS_S3_BUCKET_NAME` back up

```suggestion
    # default_s3_settings
    AWS_S3_BUCKET_NAME: str = 'Deprecated'
    AWS_S3_ADDRESSING_STYLE: str = "auto"
    AWS_S3_ENDPOINTS: dict = field(
        default_factory=lambda: {
            's3': Endpoints(),
        },
    )
    AWS_S3_KEY_PREFIX: str = ""
    AWS_S3_MAX_AGE_SECONDS: int = 60 * 60  # 1 hours.
    AWS_S3_REDUCED_REDUNDANCY: bool = False
    AWS_S3_CONTENT_DISPOSITION: str = ""
    AWS_S3_CONTENT_LANGUAGE: str = ""
    AWS_S3_METADATA: dict = field(default_factory=dict)
    AWS_S3_ENCRYPT_KEY: bool = False
    AWS_S3_KMS_ENCRYPTION_KEY_ID: str = ""
    AWS_S3_GZIP: bool = True
    AWS_S3_SIGNATURE_VERSION: str = "s3v4"
    AWS_S3_FILE_OVERWRITE: bool = False
    AWS_S3_USE_THREADS: bool = True
    AWS_S3_MAX_POOL_CONNECTIONS: int = 10
    AWS_S3_CONNECT_TIMEOUT: int = 60  # 60 seconds
    AWS_S3_READ_ONLY: bool = False
```

File: django_s3_storage/storage.py [hunk 4]
```diff
@@ -92,34 +123,81 @@ def open(self, mode="rb"):
...
+        }
+
+        # Combine settings from kwargs and Django. kwargs settings should have higher override Django settings.
+        merged_settings = d_settings | k_settings
+
+        return cls(**merged_settings)
 
-    """
-    Thread-local connection manager.
+    def __post_init__(self):
```
Comment (oskarerikmartinandersson) [inline django_s3_storage/storage.py]: Do we this method?

File: django_s3_storage/storage.py [hunk 5]
```diff
@@ -129,137 +207,103 @@ class S3Storage(Storage):
...
-        "AWS_S3_KMS_ENCRYPTION_KEY_ID": "",
-        "AWS_S3_GZIP": True,
-        "AWS_S3_SIGNATURE_VERSION": "s3v4",
-        "AWS_S3_FILE_OVERWRITE": False,
-        "AWS_S3_USE_THREADS": True,
-        "AWS_S3_MAX_POOL_CONNECTIONS": 10,
-        "AWS_S3_CONNECT_TIMEOUT": 60  # 60 seconds
-    }
-
     s3_settings_suffix = ""
```
Comment (oskarerikmartinandersson) [inline django_s3_storage/storage.py]: Should this attribute be removed? I can't find any reference to it anymore?

File: django_s3_storage/storage.py [hunk 6]
```diff
@@ -369,40 +426,64 @@ def get_available_name(self, name, max_length=None):
...
+    def validate_s3_path(self, s3_path: str):
+        url_split = urlsplit(s3_path)
+        ok_schema = (
+            url_split.scheme in self._clients.keys()
+            or url_split.scheme in self._clients_presigning.keys()
+        )
+        ok_bucket = url_split.netloc != ''
+        if not ok_schema or not ok_bucket:
+            raise NotAS3Url(
+                f"The filename {s3_path} is not a full S3 URL. Have you forgotten to set 'upload_to' on the FileField?"
```
Comment (oskarerikmartinandersson) [inline django_s3_storage/storage.py]: Is this error message a 1928 specific implementation or is this general to the overall changes you have made?

File: django_s3_storage/storage.py [hunk 7]
```diff
@@ -446,24 +529,15 @@ def size(self, name):
             return meta["ContentLength"]
 
     def url(self, name, extra_params=None, client_method="get_object"):
-        # Use a public URL, if specified.
-        if self.settings.AWS_S3_PUBLIC_URL:
-            if extra_params or client_method != "get_object":
-                raise ValueError("Use of extra_params or client_method is not allowed with AWS_S3_PUBLIC_URL")
-            return urljoin(self.settings.AWS_S3_PUBLIC_URL, filepath_to_uri(name))
         # Otherwise, generate the URL.
```
Comment (oskarerikmartinandersson) [inline django_s3_storage/storage.py]: Comment doesn't work after above code was removed

File: tests/django_s3_storage_test/tests.py [hunk 8]
```diff
@@ -15,20 +14,29 @@
...
-    def tearDown(self):
-        # clean up the dir
-        for entry in default_storage.listdir(""):
-            default_storage.delete("/".join(entry))
+class TestS3Storage(SimpleTestCase):
+    # def tearDown(self):
+    #     # clean up the dir
+    #     for entry in default_storage.listdir(helper_old_to_new_name("")):
+    #         print(entry)
+    #         # default_storage.delete("/".join(entry))
```
Comment (oskarerikmartinandersson) [inline tests/django_s3_storage_test/tests.py]: Remove unused code.

```suggestion
```

File: tests/django_s3_storage_test/tests.py [hunk 9]
```diff
@@ -15,20 +14,29 @@
...
 
     # Helpers.
 
     @contextmanager
-    def save_file(self, name="foo.txt", content=b"foo", storage=default_storage):
+    def save_file(
+        self,
+        name=f"s3://1928-django-s3-storage-test/foo.txt",
```
Comment (oskarerikmartinandersson) [inline tests/django_s3_storage_test/tests.py]: Unnecessary f-string

```suggestion
        name="s3://1928-django-s3-storage-test/foo.txt",
```

File: tests/django_s3_storage_test/tests.py
Comment (oskarerikmartinandersson) [inline tests/django_s3_storage_test/tests.py:1]: Unused imports, should be:
```python
import time
from contextlib import contextmanager
from datetime import timedelta
from urllib.parse import urlsplit, urlunsplit

import requests
from django.core.exceptions import ImproperlyConfigured
from django.core.files.base import ContentFile
from django.core.files.storage import default_storage
from django.test import SimpleTestCase
from django.utils import timezone
from django.utils.timezone import is_naive, make_naive, utc
```

File: django_s3_storage/storage.py (hunk 5 above)
Comment (hultin) [reply django_s3_storage/storage.py]: Yes, leftover from StaticS3Storage and ManifestStaticS3Storage. Removed.

File: django_s3_storage/storage.py (hunk 6 above)
Comment (hultin) [reply django_s3_storage/storage.py]: It is to the general changes.

File: django_s3_storage/storage.py (hunk 4 above)
Comment (hultin) [reply django_s3_storage/storage.py]: Leftover. Removed

File: tests/django_s3_storage_test/settings.py (hunk 2 above)
Comment (hultin) [reply tests/django_s3_storage_test/settings.py]: Nope, I was for debugging. Reverted.

File: tests/django_s3_storage_test/settings.py (hunk 1 above)
Comment (hultin) [reply tests/django_s3_storage_test/settings.py]: It is useful during debugging (duh!) Django. More details on error pages and better traces on exceptions. But no impact on django-s3-storage itself. 

I don't think we needed anymore for now, so I will remove it.

**Review by oskarerikmartinandersson — APPROVED** [review]
Looks good to me, one question though have we lost some commit on this branch compared to the `v1928.15.dev2` tag? Should we re-apply the message in the `RuntimeError`?

image

**Pull request merged**